presto-main/src/main/java/com/facebook/presto/operator/MapBlockEncodingBuffers.java

```diff
@@ -0,0 +1,371 @@
+    private int[] keyOffsetsOffsets;
+    private int[] valueOffsetsOffsets;
+
+    // The current incoming MapBlock is converted into ColumnarMap
+    private ColumnarMap columnarMap;
+
+    private final BlockEncodingBuffers keyBuffers;
+    private final BlockEncodingBuffers valueBuffers;
+
+    public MapBlockEncodingBuffers(DecodedBlockNode decodedBlockNode, int initialPositionCount)
+    {
+        super(initialPositionCount);
+
+        keyBuffers = createBlockEncodingBuffers(decodedBlockNode.getChildren().get(0), initialPositionCount);
+        valueBuffers = createBlockEncodingBuffers(decodedBlockNode.getChildren().get(1), initialPositionCount);
+    }
+
+    @Override
+    public void resetBuffers()
+    {
+        bufferedPositionCount = 0;
+        offsetsBufferIndex = 0;
+        lastOffset = 0;
+        hashTableBufferIndex = 0;
+        noHashTables = false;
+        resetNullsBuffer();
+
+        keyBuffers.resetBuffers();
+        valueBuffers.resetBuffers();
+    }
+
+    @Override
+    protected void setupDecodedBlockAndMapPositions(DecodedBlockNode decodedBlockNode)
+    {
+        requireNonNull(decodedBlockNode, "decodedBlockNode is null");
+
+        decodedBlockNode = mapPositions(decodedBlockNode);
+        columnarMap = (ColumnarMap) decodedBlockNode.getDecodedBlock();
+        decodedBlock = columnarMap.getNullCheckBlock();
+
+        populateNestedPositions();
+
+        keyBuffers.setupDecodedBlockAndMapPositions(decodedBlockNode.getChildren().get(0));
+        valueBuffers.setupDecodedBlockAndMapPositions(decodedBlockNode.getChildren().get(1));
+    }
+
+    @Override
+    public void accumulateRowSizes(int[] rowSizes)
+    {
+        for (int i = 0; i < positionCount; i++) {
+            rowSizes[i] += POSITION_SIZE;
+        }
+
+        ensureOffsetsOffsetsCapacity(positionCount);
+        System.arraycopy(offsets, 0, keyOffsetsOffsets, 0, positionCount + 1);
+        System.arraycopy(offsets, 0, valueOffsetsOffsets, 0, positionCount + 1);
+
+        keyBuffers.accumulateRowSizes(keyOffsetsOffsets, positionCount, rowSizes);
+        valueBuffers.accumulateRowSizes(valueOffsetsOffsets, positionCount, rowSizes);
+    }
+
+    @Override
+    protected void accumulateRowSizes(int[] offsetsOffsets, int positionCount, int[] rowSizes)
+    {
+        // The nested level positionCount could be 0.
+        if (this.positionCount == 0) {
+            return;
+        }
+
+        int lastOffset = offsetsOffsets[0];
+        for (int i = 0; i < positionCount; i++) {
+            int offset = offsetsOffsets[i + 1];
+            rowSizes[i] += POSITION_SIZE * (offset - lastOffset);
+            lastOffset = offset;
+            offsetsOffsets[i + 1] = offsets[offset];
+        }
+
+        // offsetsOffsets might be modified by the next level. Save it for the valueBuffers first.
+        ensureOffsetsOffsetsCapacity(positionCount);
+        System.arraycopy(offsetsOffsets, 0, valueOffsetsOffsets, 0, positionCount + 1);
```

Comment:
This logic should match the logic in the other `accumulateRowSizes`. As written, valueOffsetsOffsets is not used. I'd expect some test to fail, unless this code path is not covered. Could you make sure this code path is covered?

Reply:
Added more tests in TestBlockEncodingBuffers.testMap() to trigger the error. It requires three level of nested maps.